src/main/java/com/allenai/ml/optimize/QuasiNewton.java

```diff
@@ -0,0 +1,96 @@
+package com.allenai.ml.optimize;
+
+import com.allenai.ml.linalg.Vector;
+import com.gs.collections.api.tuple.Pair;
+import com.gs.collections.impl.tuple.Tuples;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Interface for how to approximate `H^{-1} dir`, where `H^{-1}` is an approximation for the inverse hessian and
+ * dir is the current search direction. If we use `H^{-1}` as the identity matrix, you get standard gradient descent.
+ * If you remember some history, you get LBFGS.
+ */
+@FunctionalInterface
+public interface QuasiNewton {
+
+    /**
+     * @return Approximation of inverse hessian matrix times `dir`
+     */
+    Vector implictMultiply(Vector dir);
+
+    /**
+     * Update the approximation
+     * @param xDelta The delta from the last guess
+     * @param gradDelta The delta from the function gradient at last guess
+     */
+    default void update(Vector xDelta, Vector gradDelta) {
+        // intentional no-op
+    }
+
+    /**
+     * Approximate inverse hessian with the identity
+     */
+    static QuasiNewton gradientDescent() {
+        return dir -> dir;
+    }
+
+    /**
+     * LBFGS inverse hessian approximation
+     */
+    static QuasiNewton lbfgs(final int maxHistorySize) {
+        return new QuasiNewton() {
+            // Store (xDelta, gradDelta) pairs
+            private final List<Pair<Vector, Vector>> history = new ArrayList<>();
+
+            private double initialScale() {
+                if (history.isEmpty()) {
+                    return 1.0;
+                }
+                Vector lastInputDiff = history.get(0).getOne();
+                Vector lastGradDiff = history.get(0).getTwo();
+                double numer = lastGradDiff.dotProduct(lastInputDiff);
+                double denom = lastGradDiff.l2NormSquared();
+                if (denom == 0.0) {
+                    throw new RuntimeException("Can't have zero curvature");
```

Comment:
The inverse curvature estimate is undefined (`numer/denom` below), it should be an impossible state in the context of optimization because modulo numerical issues, you can only have this when you're gradient is zero (the l2 norm is zero) in which case you already converged. It should probably be an `ImpossibleStateException`